lib/browser/source.js

```diff
@@ -11,7 +11,95 @@ class FileSource {
   close() {}
 }
 
-export function getSource(input) {
+class StreamSource {
+  constructor(reader, chunkSize) {
+    // Ensure that chunkSize is an integer and not something else or Infinity.
+    chunkSize = +chunkSize;
+    if (!isFinite(chunkSize)) {
+      throw new Error("cannot create source for stream without a finite value for the `chunkSize` option");
+    }
+    this._chunkSize = chunkSize;
+
+    this._buffer = undefined;
+    this._bufferOffset = 0;
+    this._reader = reader;
+    this._done = false;
+  }
+
+  slice(start, end, callback) {
+    if (start < this._bufferOffset) {
+      throw new Error("requested data is before the reader's current offset");
+    }
+
+    return this._readUntilEnoughDataOrDone(start, end, callback);
+  }
+
+  _readUntilEnoughDataOrDone(start, end, callback) {
+    const hasEnoughData = end <= this._bufferOffset + len(this._buffer);
+    if (this._done || hasEnoughData) {
+      callback(this._getDataFromBuffer(start, end));
+      return;
+    }
+    this._reader.read().then(({ value, done }) => {
+      if (done) {
+        this._done = true;
+      } else if (this._buffer === undefined){
+        this._buffer = value;
+      } else {
+        this._buffer = concat(this._buffer, value);
+      }
+
+      this._readUntilEnoughDataOrDone(start, end, callback);
+    });
+  }
+
+  _getDataFromBuffer(start, end) {
+    const hasAllDataBeenRead = start >= this._bufferOffset + len(this._buffer);
+    if (this._done && hasAllDataBeenRead) {
+      return null;
+    }
+    const bufferToReturn = this._buffer.slice(
+      start - this._bufferOffset,
+      end - this._bufferOffset
+    );
+    const chunkSize = end - start;
+    if (len(this._buffer) >= 2 * chunkSize) {
```

Comment:
Would you mind explaining this condition? Maybe that's also worth to be added as comment.

Reply:
Embarrassingly, this code was wrong! Attempting to write a comment made me realize my error 😬

The thinking was that we can't remove code from the buffer until we know it has been successfully uploaded. I tried to achieve this by always keeping twice the chunkSize to ensure we had the last chunk of data for resending. This approach is wrong.

The size of the buffer doesn't matter, and instead we can only delete data that is _before_ the currently requested chunk. e.g. If the uploader is requesting bytes 1000-1100, we know that bytes < 1000 have been successfully uploaded.

I fixed the code, added a test, and described the behavior in a comment.